src/coalton.lisp

```diff
@@ -115,33 +115,105 @@ in FORMS that begin with that operator."
+                         ((coalton:define defines))
+                         ((coalton:define-type type-defines))
+                         ((coalton:define-class class-defines))
+                         ((coalton:define-instance instance-defines))
+                         ((repr-table repr-table))
+                       &allow-other-keys)
+      (collect-toplevel-forms toplevel-forms)
 
+    (multiple-value-bind (defined-types env)
+        (process-toplevel-type-definitions type-defines repr-table env)
+
+      ;; Class definitions must be checked after types are defined
+      ;; but before values are typechecked.
+
+      (multiple-value-bind (classes env)
+          (parse-class-definitions class-defines env)
+
+        ;; Methods need to be added to the environment before we can
+        ;; check value types.
+        (setf env (predeclare-toplevel-instance-definitions instance-defines package env))
+
+        (let ((declared-types (process-toplevel-declarations declares env)))
+          (multiple-value-bind (env toplevel-bindings)
+              (process-toplevel-value-definitions defines declared-types package env)
+
+            ;; Methods must be typechecker after the types of values
+            ;; are determined since instances may reference them.
+            (let* ((instance-definitions
+                     (process-toplevel-instance-definitions
+                      instance-defines
+                      package
+                      env))
+
+                   (translation-unit
+                     (make-translation-unit
+                      :types defined-types
+                      :definitions toplevel-bindings
+                      :instances instance-definitions
+                      :classes classes
+                      :package package)))
+
+              (multiple-value-bind (program env)
+                  (coalton-impl/codegen:compile-translation-unit
+                   translation-unit
+                   env)
+
+                (values
+                 (if *coalton-skip-update*
+                     program
+                     `(progn
+                        (eval-when (:load-toplevel)
+                          (unless (eq (coalton-release-p) ,(coalton-release-p))
+                            ,(if (coalton-release-p)
+                                `(error "~A was compiled in release mode but loaded in development." ,(or *compile-file-pathname* *load-truename*))
+                                `(error "~A was compiled in development mode but loaded in release." ,(or *compile-file-pathname* *load-truename*)))))
+                        ,(coalton-impl/typechecker::generate-diff
+                          translation-unit
+                          env
+                          '*global-environment*)
+                        ,program))
+                 env)))))))))
+
+;;;
 ;;; Coalton Macros
-(defmacro coalton:coalton-toplevel (&body toplevel-forms)
+;;;
+
+(in-package #:coalton)
```

Comment:
I don't understand why you're not just in-package coalton-impl and qualifying them coalton symbols as done already